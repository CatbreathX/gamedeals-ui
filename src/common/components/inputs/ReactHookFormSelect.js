import { FormControl, InputLabel, Select } from '@mui/material';
import PropTypes from 'prop-types';
import { Controller } from 'react-hook-form';

export const ReactHookFormSelect = ({
  name,
  label,
  control,
  defaultValue,
  children,
  ...props
}) => {
  const labelId = `${name}-label`;

  return (
    <FormControl {...props}>
      <InputLabel id={labelId}>{label}</InputLabel>
      <Controller
        render={({ field }) => (
          <Select
            {...field}
            name={name}
            labelId={labelId}
            label={label}
          >
            {children}
          </Select>
        )}
        name={name}
        control={control}
        defaultValue={defaultValue}
      />
    </FormControl>
  );
};

ReactHookFormSelect.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  control: PropTypes.object.isRequired,
  defaultValue: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
};
