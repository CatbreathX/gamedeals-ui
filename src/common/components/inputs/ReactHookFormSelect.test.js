import { MenuItem } from '@mui/material';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ReactHookFormSelect } from 'common/components/inputs/ReactHookFormSelect';
import { FormProvider, useForm } from 'react-hook-form';

describe('ReactHookFormSelect', () => {
  test('should display correct label', () => {
    mount();
    expect(screen.getByLabelText('Select Unit Test')).toBeInTheDocument();
  });

  test('should render all options', () => {
    mount();

    const button = screen.getByRole('button');
    userEvent.type(button, '{arrowdown}');

    const actualOptions = screen.getAllByRole('option');
    expect(actualOptions).toHaveLength(selectOptions.length);

    selectOptions.forEach((option) => {
      expect(screen.getByRole('option', { name: option.text })).toBeInTheDocument();
    });
  });

  test('should select default value', () => {
    mount('2');
    expect(screen.getByRole('button')).toHaveTextContent('Two');
  });

  test('should update with selected value', () => {
    mount();

    const button = screen.getByRole('button');
    userEvent.type(button, '{arrowdown}');
    userEvent.type(screen.getAllByRole('option')[0], '{enter}');
    expect(button).toHaveTextContent('One');
  });
});

const mount = (defaultValue = '') => {
  render(
    <Select defaultValue={defaultValue} />,
  );
};

// eslint-disable-next-line react/prop-types
const Select = ({ defaultValue }) => {
  const methods = useForm({
    defaultValues: {
      title: 'Elite',
    },
  });

  return (
    <FormProvider {...methods}>
      <ReactHookFormSelect
        name="test"
        label="Select Unit Test"
        control={methods.control}
        defaultValue={defaultValue}
      >
        {selectOptions.map((option) => (
          <MenuItem key={option.value} value={option.value}>{option.text}</MenuItem>))}
      </ReactHookFormSelect>
    </FormProvider>
  );
};

const selectOptions = [
  {
    value: '1',
    text: 'One',
  },
  {
    value: '2',
    text: 'Two',
  },
  {
    value: '3',
    text: 'Three',
  },
];
