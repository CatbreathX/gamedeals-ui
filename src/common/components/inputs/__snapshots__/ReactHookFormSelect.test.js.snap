// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactHookFormSelect should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      id="test-label"
    >
      Select Unit Test
    </label>
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="test-label mui-component-select-test"
        class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="mui-component-select-test"
        role="button"
        tabindex="0"
      >
        <span
          class="notranslate"
        >
          â€‹
        </span>
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        name="test"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-1ftyaf0"
        >
          <span>
            Select Unit Test
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
